amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,422 @@
+//! Contains the `Error` type and company as used by Amethyst.
+//!
+//! **Note:** This type is not intended to be used outside of Amethyst.
+//! If you are integrating a crate from amethyst to use this, it is recommended that you treat this
+//! type as an opaque [`std::error::Error`].
+//!
+//! [`std::error::Error`]: https://doc.rust-lang.org/std/error/trait.Error.html
+
+// Parts copied from failure:
+// https://github.com/rust-lang-nursery/failure
+
+#![warn(missing_docs, rust_2018_idioms, rust_2018_compatibility)]
+
+#[cfg(all(feature = "backtrace", feature = "std"))]
+extern crate backtrace;
+
+use self::internal::Backtrace;
+use std::borrow::Cow;
+use std::error;
+use std::fmt;
+use std::result;
+
+/// Internal parts of `Error`.
+#[derive(Debug)]
+struct Inner<T: ?Sized> {
+    source: Option<Box<Error>>,
+    #[cfg(all(feature = "backtrace", feature = "std"))]
+    backtrace: Option<Backtrace>,
+    error: T,
+}
+
+/// The error type used by Amethyst.
+///
+/// Wraps error diagnostics like messages and other errors, and keeps track of causal chains and
+/// backtraces.
+pub struct Error {
+    inner: Box<Inner<dyn error::Error + Send + Sync>>,
```

Comment:
This way `Error` size is two pointer.
Considering that wrapped errors should be rare it make sense to limit its side to one pointer.
Even at the cost of additional allocation e.g. `Box<Box<UnsizedType>>`.
Because this cost only applied to error path, while happy path gets smaller return type.

Reply:
I'm not convinced. A `2*usize` return value is already tiny and doesn't automatically hold its own weight compared to the extra indirection another pointer would for me.

Reply:
Return type size can make significant difference in generated instructions. And additional indirection occur only in case of error, when, as you stated, performance is not of concern.

But we probably should benchmark it first.

Reply:
Since this doesn't affect the API I suggest resolving this discussion. We can experiment (and change) this without much effort if we find the need to in the future.